src/target/riscv/riscv.c

```diff
@@ -598,90 +606,174 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
+		if (ret != ERROR_OK)
+			continue;
+		ret = set_trigger(target, idx + 1, trig2_info.tdata1, trig2_info.tdata2,
+						  trig2_info.tdata1_ignore_mask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			r->trigger_unique_id[idx + 1] = trigger->unique_id;
+			return ERROR_OK;
+		} else
+			set_trigger(target, idx, 0, 0,
+						0); /* Undo the setting of the previous trigger*/
 	}
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_LT);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	tdata2 = trigger->address + trigger->length;
-	ret = set_trigger(target, idx + 1, tdata1, tdata2, 0);
-	if (ret != ERROR_OK) {
-		/* Undo the setting of the previous trigger*/
-		set_trigger(target, idx, 0, 0, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
-		return ret;
+	return ret;
+}
+
+static int try_trigger_request_match_with_fallbacks(
+	struct target *target, struct trigger *trigger, int trigger_type,
+	struct trigger_request_info napot_trigger_info,
+	struct trigger_request_info trig1_ge_info,
+	struct trigger_request_info trig2_lt_info,
+	struct trigger_request_info trig1_lt_info,
+	struct trigger_request_info trig2_ge_info,
+	struct trigger_request_info trig_eq_info)
+{
+	int ret = ERROR_OK;
+
+	if (!(trigger->execute || trigger->length == 1)) {
+		if (can_use_napot_match(trigger)) {
+			napot_trigger_info.tdata2 =
+				trigger->address | ((trigger->length - 1) >> 1);
+			LOG_TARGET_DEBUG(target, "trying t request NAPOT match");
+			ret = try_trigger_request_match(target, trigger, trigger_type,
+											napot_trigger_info);
+			if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+				return ret;
+		}
+		LOG_TARGET_DEBUG(target, "trying t request GE/LT match");
+		ret = try_chained_trigger_request_match(target, trigger, trigger_type,
+												trig1_ge_info, trig2_lt_info);
+		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			return ret;
+
+		LOG_TARGET_DEBUG(target, "trying t request LT/GE match");
+		ret = try_chained_trigger_request_match(target, trigger, trigger_type,
+												trig1_lt_info, trig2_ge_info);
+		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			return ret;
 	}
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	r->trigger_unique_id[idx + 1] = trigger->unique_id;
-	return ERROR_OK;
+	LOG_TARGET_DEBUG(target, "trying t request EQUAL match");
+	return try_trigger_request_match(target, trigger, trigger_type,
+									 trig_eq_info);
+}
 
-MATCH_EQUAL:
-	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
-	if (ret != ERROR_OK)
-		return ret;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_EQUAL);
+static int maybe_add_trigger_t2(struct target *target, struct trigger *trigger)
+{
+	RISCV_INFO(r);
+
+	struct trigger_request_info common = {
+		.tdata1 = 0, .tdata2 = trigger->address, .tdata1_ignore_mask = 0};
```

Comment:
Please, format the code like this - each assignment on a separate line:

```suggestion
	struct trigger_request_info common = {
		.tdata1 = 0,
		.tdata2 = trigger->address,
		.tdata1_ignore_mask = 0};
```

Reply:
Addressed.